### ISSUE TITLE
Support for compressed_texture_s3tc with webgl

### ISSUE THREAD

**Issue by JiDW** [pixijs/pixi.js#1343]
Feature request, following this discussion : http://www.html5gamedevs.com/topic/10086-whats-the-current-status-of-pixis-support-for-compressed-textures/

As a reminder, from OpenGL wiki : 

"S3TC is a technique for compressing images for use as textures. Standard image compression techniques like JPEG and PNG can achieve greater compression ratios than S3TC. However, S3TC is designed to be implemented in high-performance hardware. JPEG and PNG decompress images all-at-once, while S3TC allows specific sections of the image to be decompressed independently." 

It is a very useful feature for mobile development, where you care a lot more about memory management than download size.

**Comment by englercj** [comment]
I looked into this a little bit, and I don't think there is anything stopping us from adding support for it. Only thing is we want to get the v3 rewrite done before we get back to feature development so it may be a few weeks before you see this.

**Comment by mattdesl** [comment]
Just to chime in, I used toji's [texture utils](https://github.com/toji/webgl-texture-utils) to get ST3C working in a library of mine ([see old code](https://github.com/mattdesl/kami-demos/tree/master/src/s3tc)). 

toji's code is not easily "consumable" with bower or npm -- this means we need to keep an eye on his repo and copy-paste any patches or bug fixes.. not ideal. toji also does not maintain the repo anymore. And given that it's a pretty huge dependency, this is the sort of thing that would be ideal as a third-party solution.

This is a perfect area for "userland" growth. The generic Pixi API allows for custom texture types, and then people can build modules that target this use-case. :+1:

**Comment by englercj** [comment]
:+1: To what Matt says, with the extensible v3 API we are really looking for the community to step up and build userland modules.

**Comment by JiDW** [comment]
Hello there,

We're working on this for our project, but we don't know how to integrate this into pixi, being a plugin (seems hard to do) or a core feature.

The first thing to do is to be able to call gl.compressedTexImage2D() instead of gl.texImage2D in WebGLRenderer.prototype.updateTexture when the format of the base texture is a compressed format.

The second thing is to make ressourceLoader deals with compressed format files

The last thing is to deal with compatibility issues. This can be done outside of PIXI or thanks to a .utils function or RessourceLoader as is require a bit of code (check http://toji.github.io/texture-tester/ )

There is at this time 4 compressed texture formats and we need to be compatible with all of them, as it is an hardware compatibility, not a software compatibility. Each GPU model can decode a single (or two) format.

Would you have any insight about how to intregrate this feature inside PIXI ?

**Comment by JiDW** [comment]
Hi again.

Trying to hype people here (and hopefully get some answers) : we managed to deal with compressed textures in our own engine (we use a small 3D engine on top of PIXI) and the results are super good !

Unfortunatly, after taking the time to look deeper into the tasks needed to make PIXI deals with compressed textures, I'm not confident enough to modify PIXI in its core functions... sorry :/

However, I do know what already works and what needs to be done.

1- PIXI needs to be able to use an Int32Array as a BaseTexture.source for compressed textures. The ressourceloader needs to be modified in order to load raw datas and I contacted Chad about it on github.
2- (copy/paste)To use gl.compressedTexImage2D() instead of gl.texImage2D in WebGLRenderer.prototype.updateTexture when the format of the base texture is a compressed format.

Why is that important and why am I pushy about it ?

Well, even with the least compressed format, a DDS/DXT5 textures takes 4x less gpu memory than a png or jpg.

It means so much for mobile games, this is huge !

And even for desktop, a compressed texture should take less time to bind. PIXI being a 2D engine and, for good reasons, not using a depth buffer, binding textures can often be a performance issue in games with a lot of assets.

Regards.
